[1/2] `WCProductModel` to Room: working app

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->
- #13935
- #13943 

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCProductModel` from WellSql to Room ORM. 

- `WCProductModel` will be now `data class` and its field will be immutable.
- I didn't change the name of the class to a more suitable `ProductEntity` to not overshadow the changes. We can do this in a separate PR.
- Several `runBlocking` calls were introduced, but they're only here to **highlight** where we execute database calls on the main thread. All DAO method are either `suspend` or return `Flow`.
- The real line diff of this PR is (**+856 / -1088**). This excludes JSON schema file.
- **`ProductsDao` and `ProductSqlUtils` are now `internal`** classes of `fluxc-plugin`
- This PR is not meant to pass CI checks. See #13943 which updates unit tests to the changes introduces in this PR.
- `rowsAffected` features were removed from the app. Room doesn't support this nicely, and I realized it's not really necessary to have it.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Please use #13943 for tests.


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1665" alt="image" src="https://github.com/user-attachments/assets/be3369e6-0d90-4642-b528-306bbd181568" />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Remove unused methods

wzieba committed:
Migrate `observeProductsCount`.

Don't migrate `debounce`: this is rather a UI requirement, shouldn't be a part of data layer

wzieba committed:
Migrate `observeProducts`.

Without sorting yet

wzieba committed:
Apply sorting for observing products

This is the initial idea, might be changed to enums with type converters or maybe something else

wzieba committed:
Migrate `observeBundledProducts`

Simplify how bundled products are retrieved

wzieba committed:
Migrate `getProductByRemoteId`

wzieba committed:
Migrate `insertOrUpdateProduct`

Replace with Room's `@Upsert`

wzieba committed:
Bring back `rowsAffected` to `upsertProducts`

wzieba committed:
Remove unused methods

wzieba committed:
Make `getProducts` suspended

Update code to support this. `runBlocking` should be addressed in future work

wzieba committed:
Migrate `getProductsByRemoteIds`

And replace it with Room's DAO Query

wzieba committed:
Remove `getProductCountByRemoteIds`

Another method only for counting is not really needed: I think performance gain of checking size in SQL is not noticeable in our case

wzieba committed:
Remove `getVirtualProductCountByRemoteIds`

Merge it with `getProducts` query

wzieba committed:
Replace WellSql in searchable "getProducts"

wzieba committed:
Apply search logic in Kotlin, instead of SQL query

I didn't want to make such complex logic in raw SQL. I think performance-wise it won't be a noticeable difference

wzieba committed:
Remote unused methods

wzieba committed:
Migrate `getProductExistsByRemoteId` to Room

wzieba committed:
Migrate `getProductExistsBySku` to Room

wzieba committed:
Migrate `deleteProductsForSite` to Room

wzieba committed:
Migrate `deleteProduct` to Room

wzieba committed:
Make the app compile: adjust codebase to new WCProductModel, add necessary code to create new Room table

wzieba committed:
Fix incorrect filtering logic on `ProductsDao#getProducts`

wzieba committed:
Use `runBlocking` close to the source

Initially, I wanted to move `runBlocking` wrappers away from source, to the presentation layer, as this would be easier to fix and manage. Unfortunately, it's not possible as tests rely on synchronous nature of these calls. So for now, I'm moving the `runBlocking` wrappers as close to the source as possible

wzieba committed:
Add WellSql migration to drop `WCProductModel` table in it

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProductsDao</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'trunk' into migrate_products_to_room

wzieba committed:
Move sorting from SQL

- the approach with dynamic sql query didn't work: Room seems to not allow dynamic names of parameters to probably prevent any SQL-injection vectors
- the difference in performance shouldn't be noticeable: list of products are not that huge

wzieba committed:
Remove `observeProductsCount` method

After rethinking, here again I think using SQL is premature optimization that is not noticeable by the end user, but complicates the `ProductsDao`

wzieba committed:
Apply static code analysis fixes

wzieba committed:
Remove `rowsAffected` from `OnProductChanged`

This property wasn't used in any meaningful way.

wzieba committed:
Merge branch 'trunk' into migrate_products_to_room

wzieba committed:
Remove `rowsAffected` from `OnProductCreated`, `OnProductUpdated`, `OnProductImagesChanged`

This property wasn't used in any meaningful way.

wzieba committed:
Make `ProductsDao` and `ProductSqlUtils` `internal` for fluxc-plugin module

The data should be access only via `WCProductStore`

wzieba committed:
Rename `WCProductModel` to `ProductEntity` Room table

To keep it consistent with other tables.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/details/ProductDetailRepository.kt:
```diff
@@ -328,7 +329,7 @@ class ProductDetailRepository @Inject constructor(
     }
 
-    fun isSkuAvailableLocally(sku: String) = !productStore.isProductExists(selectedSite.get(), sku)
+    fun isSkuAvailableLocally(sku: String) = runBlocking { !productStore.isProductExists(selectedSite.get(), sku) }
```
Yes sure, like in other places in this PR I intentionally block main thread using `runBlocking`. The reason is that it was like this before: database calls were performed on main thread synchronously. Changing this is out of scope of this PR as it will affect flow of events in the app and has to be carefully handled case-by-case.

wzieba committed:
Make upsert products methods `suspend`

wzieba committed:
Make `excludedProductIds` non-nullable

Room doesn't work well with nullable lists in this context. This commit backports ad0d43b and f192a53 commits from another branch.

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt:
```diff
@@ -0,0 +1,114 @@
+            AND (:stockStatus IS NULL OR stockStatus = :stockStatus)
+            AND (:type IS NULL OR type = :type)
+            AND (:category IS NULL OR categories LIKE '%' || :category || '%')
+            AND (:excludeSampleProducts IS NULL OR isSampleProduct = :excludeSampleProducts)
```
I don't think this condition is correct, because `excludeSampleProducts` is not nullable, and `isSampleProduct` is not equivalent to `excludeSampleProducts`.

I think something like the following would work, WDYT?
```suggestion
            AND (:excludeSampleProducts = 0 OR isSampleProduct = 0)
```

wzieba replied:
Very good point, thanks. Addressed in a131423c1fc0f11ffcb7a441474d2647cc45682c

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt:
```diff
@@ -1034,23 +1047,33 @@ class WCProductStore @Inject constructor(
-        filterOptions = filterOptions,
-        excludeSampleProducts = excludeSampleProducts
-    )
+    ): Flow<Long> = productsDao.observeProducts(
```
Loading the whole table into memory to count could create issues, let's try to count in the DB itself.

For context, we have stores with thousands of products (you can check the counts of products of some stores by using this SQL query 3796a-pb in Superset).

wzieba replied:
Thank you: as I mentioned in another comment, I didn't suspect stores could have thousands of products. I've applied the suggestion 6b219fa63a3e9b71b48814a4a807943ea90c175a. I like it, as it also allowed to remove some of the unnecessary parameters from the SQL query (like `LIMIT` when we count products).

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt:
```diff
@@ -122,299 +75,125 @@ object ProductSqlUtils {
+            if (searchQuery.isNullOrBlank()) {
+                true
+            } else {
+                when (skuSearchOptions) {
```
Doing the search in the app has two issues:

- Less performant, and as shared above we have stores with thousands of products.
- More importantly, it could lead to wrong results, because the function accepts a `limit` argument, which means that the DB could return just a subset of the data and we'll search just against it.

Why the search wasn't handled in the DB itself? I think we can have different Dao functions for the different search options, WDYT?

wzieba replied:
Very good point on the `limit`. I decided to introduce following approach: 31c8f6b9443706af44c55fcdebde2e972f7c50ae

I think we don't need a "fat" `WCProductModel#getProducts` method with possibilities like `limit` or applying filters when searching. It's not used like this in the app. 

To improve clarity, I decided to split this method and create `WCProductModel#searchCachedProducts` with much smaller and targeted DAO methods.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/grouped/GroupedProductListViewModel.kt:
```diff
@@ -121,17 +121,19 @@ class GroupedProductListViewModel @Inject constructor(
+                )
+                if (productsInDb.isNotEmpty()) {
+                    _productList.value = productsInDb
+                    productListViewState = productListViewState.copy(isSkeletonShown = false)
```
If I understand correctly, this code can be executed after the one in the init

```
        productListViewState = productListViewState.copy(
            isEmptyViewShown = navArgs.productIds.isEmpty() &&
                selectedProductIds.isEmpty()
        )
```

I don't see any issues with that, but I'm just thinking out loud that there is a change in the behavior introduced.

wzieba replied:
Thanks for pointing this out. Actually, I've reverted this change in d98312fa865af6252555afc68466a810de727f33 - we don't have to wrap this with `launch` at this moment, so there's no risk in changing the behavior here.

kidinov commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductModel.kt:
```diff
@@ -24,103 +26,109 @@ import org.wordpress.android.util.AppLog.T
+@Entity(
+    tableName = "ProductEntity",
+    indices = [Index(
+        value = ["localSiteId", "remoteId"],
```
np: AFAIR we don't need to declare indices on the primary keys, indeces will be built anyway on those fields

wzieba replied:
Thanks! Addressed in e2e53a7a393c40bc58240eff271b318bfedc36ac . I had a memory from migrating `OrderEntity` that when I didn't add index, `kapt` was throwing a warning, but it's not the case now 👍 .

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt:
```diff
@@ -122,299 +75,125 @@ object ProductSqlUtils {
+    }
 
-    private fun getSortOrder(sortType: ProductSorting) =
+    internal fun List<WCProductModel>.sort(sortType: ProductSorting): List<WCProductModel> =
```
Doing the sort on the app is problematic for the same reasons mentioned before:

- Less performant.
- The `limit` argument would lead to bugs, as the subset of items we'll get from the DB will be wrong. During my testing, this caused an issue in the Blaze card in the dashboard, as we retrieve only the two most recent products there.

One approach to achieve sorting in the DB would be something like this patch
<details>
<summary> Patch </summary>

```patch
Index: libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt
--- a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt	(revision 07324874484fa97d2514918ad6b7166af9cd2414)
+++ b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt	(date 1745345480780)
@@ -138,9 +138,9 @@
             category = filterOptions[ProductFilterOption.CATEGORY]?.let { categoryFilter(it) },
             excludeSampleProducts = excludeSampleProducts,
             limit = limit,
-            excludedProductIds = excludedProductIds
+            excludedProductIds = excludedProductIds,
+            sortType = sortType
         ).firstOrNull().orEmpty()
-            .sort(sortType)
             .filter { product ->
             if (searchQuery.isNullOrBlank()) {
                 true
@@ -172,16 +172,6 @@
         return products
     }
 
-    internal fun List<WCProductModel>.sort(sortType: ProductSorting): List<WCProductModel> =
-        when (sortType) {
-            TITLE_ASC -> sortedBy { it.name }
-            TITLE_DESC -> sortedByDescending { it.name }
-            DATE_ASC -> sortedBy { it.dateCreated }
-            DATE_DESC -> sortedByDescending { it.dateCreated }
-            POPULARITY_ASC -> sortedBy { it.totalSales }
-            POPULARITY_DESC -> sortedByDescending { it.totalSales }
-        }
-
     fun insertOrUpdateProductVariation(variation: WCProductVariationModel): Int {
         val result = WellSql.select(WCProductVariationModel::class.java)
             .where().beginGroup()
@@ -634,5 +624,4 @@
     private fun triggerCategoriesUpdateIfNeeded(affectedRows: Int) {
         if (affectedRows != 0) categoriesUpdatesTrigger.tryEmit(Unit)
     }
-
 }
Index: libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt
--- a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt	(revision 07324874484fa97d2514918ad6b7166af9cd2414)
+++ b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt	(date 1745347252720)
@@ -5,6 +5,13 @@
 import androidx.room.Upsert
 import kotlinx.coroutines.flow.Flow
 import org.wordpress.android.fluxc.model.WCProductModel
+import org.wordpress.android.fluxc.store.WCProductStore
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.DATE_ASC
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.DATE_DESC
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.POPULARITY_ASC
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.POPULARITY_DESC
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.TITLE_ASC
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.TITLE_DESC
 
 @Dao
 internal abstract class ProductsDao {
@@ -17,12 +24,18 @@
             AND (:stockStatus IS NULL OR stockStatus = :stockStatus)
             AND (:type IS NULL OR type = :type)
             AND (:category IS NULL OR categories LIKE '%' || :category || '%')
-            AND (:excludeSampleProducts IS NULL OR isSampleProduct = :excludeSampleProducts)
+            AND (:excludeSampleProducts = 0 OR isSampleProduct = 0)
             AND (remoteId NOT IN (:excludedProductIds))
+            ORDER BY
+                CASE WHEN :sortType = 'TITLE_ASC' THEN name END ASC,
+                CASE WHEN :sortType = 'TITLE_DESC' THEN name END DESC,
+                CASE WHEN :sortType = 'DATE_ASC' THEN dateCreated END ASC,
+                CASE WHEN :sortType = 'DATE_DESC' THEN dateCreated END DESC,
+                CASE WHEN :sortType = 'POPULARITY_ASC' THEN totalSales END ASC,
+                CASE WHEN :sortType = 'POPULARITY_DESC' THEN totalSales END DESC
             LIMIT CASE WHEN :limit IS NULL THEN -1 ELSE :limit END
         """
     )
-    @Suppress("LongParameterList")
     abstract fun observeProducts(
         localSiteId: Int,
         status: String?,
@@ -31,7 +44,8 @@
         category: String?,
         excludeSampleProducts: Boolean,
         limit: Int?,
-        excludedProductIds: List<Long>
+        excludedProductIds: List<Long>,
+        sortType: WCProductStore.ProductSorting
     ): Flow<List<WCProductModel>>
 
     @Query(
Index: libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt
--- a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt	(revision 07324874484fa97d2514918ad6b7166af9cd2414)
+++ b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt	(date 1745347302281)
@@ -51,12 +51,12 @@
 import org.wordpress.android.fluxc.persistence.ProductSqlUtils.getCompositeProducts
 import org.wordpress.android.fluxc.persistence.ProductSqlUtils.getProducts
 import org.wordpress.android.fluxc.persistence.ProductSqlUtils.observeBundledProducts
-import org.wordpress.android.fluxc.persistence.ProductSqlUtils.sort
 import org.wordpress.android.fluxc.persistence.ProductStorageHelper
 import org.wordpress.android.fluxc.persistence.dao.AddonsDao
 import org.wordpress.android.fluxc.persistence.dao.ProductsDao
 import org.wordpress.android.fluxc.store.WCProductStore.ProductCategorySorting.NAME_ASC
 import org.wordpress.android.fluxc.store.WCProductStore.ProductErrorType.GENERIC_ERROR
+import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.DATE_ASC
 import org.wordpress.android.fluxc.store.WCProductStore.ProductSorting.TITLE_ASC
 import org.wordpress.android.fluxc.tools.CoroutineEngine
 import org.wordpress.android.fluxc.utils.AppLogWrapper
@@ -1057,7 +1057,8 @@
             excludeSampleProducts = excludeSampleProducts,
             excludedProductIds = emptyList(),
             limit = limit,
-        ).map { it.sort(sortType) }
+            sortType = sortType
+        )
     }
 
     fun observeProductsCount(
@@ -1072,7 +1073,8 @@
         category = filterOptions[ProductFilterOption.CATEGORY]?.let { categoryFilter(it) },
         excludeSampleProducts = excludeSampleProducts,
         limit = null,
-        excludedProductIds = emptyList()
+        excludedProductIds = emptyList(),
+        sortType = DATE_ASC
     ).map { it.size.toLong() }
 
     fun observeVariations(site: SiteModel, productId: Long): Flow<List<WCProductVariationModel>> =
```
</details>

A second approach would be to simply use `RawQuery` and write the query manually.

WDYT?

wzieba replied:
Sure! I didn't consider that shops can have thousands of products, so I thought database-level sorting/filtering was not necessary. 

I addressed the suggestion in 980ccaf5dd8d2a7e9e5e36742f827c540b756855 with your patch.

I was considering `RawQuery` but I felt it was getting not very readable quickly. Actually, WellSql had a nicer way of building the query :)

hichamboushaba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt:
```diff
@@ -122,299 +75,125 @@ object ProductSqlUtils {
-                .endGroup().endWhere()
-                .exists()
-    }
+        val products = observeProducts(
```
Nit, just a personal opinion, I don't like returning a `Flow` from the DB just to get one item, besides that it has some overhead, it doesn't seem very readable.

I understand that exposing two functions with the same query in the Dao would mean some repetition, but I think we can reduce the repetition by using a constant query, like the below patch:

<details>
<summary>Patch</summary>

```patch
Index: libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt
--- a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt	(revision 07324874484fa97d2514918ad6b7166af9cd2414)
+++ b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/ProductSqlUtils.kt	(date 1745347796063)
@@ -16,7 +16,6 @@
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.MutableSharedFlow
 import kotlinx.coroutines.flow.debounce
-import kotlinx.coroutines.flow.firstOrNull
 import kotlinx.coroutines.flow.flowOn
 import kotlinx.coroutines.flow.mapLatest
 import kotlinx.coroutines.flow.onStart
@@ -130,7 +129,7 @@
         limit: Int? = null
     ): List<WCProductModel> {
 
-        val products = observeProducts(
+        val products = getProducts(
             localSiteId = site.id,
             status = filterOptions[ProductFilterOption.STATUS],
             stockStatus = filterOptions[ProductFilterOption.STOCK_STATUS],
@@ -139,35 +138,34 @@
             excludeSampleProducts = excludeSampleProducts,
             limit = limit,
             excludedProductIds = excludedProductIds
-        ).firstOrNull().orEmpty()
-            .sort(sortType)
+        ).sort(sortType)
             .filter { product ->
-            if (searchQuery.isNullOrBlank()) {
-                true
-            } else {
-                when (skuSearchOptions) {
-                    SkuSearchOptions.Disabled -> {
-                        searchQuery.let { query ->
-                            listOf(
-                                product.name,
-                                product.description,
-                                product.shortDescription
-                            ).any { it.contains(query) }
-                        }
-                    }
+                if (searchQuery.isNullOrBlank()) {
+                    true
+                } else {
+                    when (skuSearchOptions) {
+                        SkuSearchOptions.Disabled -> {
+                            searchQuery.let { query ->
+                                listOf(
+                                    product.name,
+                                    product.description,
+                                    product.shortDescription
+                                ).any { it.contains(query) }
+                            }
+                        }
 
-                    SkuSearchOptions.ExactSearch -> {
-                        product.sku == searchQuery
-                    }
+                        SkuSearchOptions.ExactSearch -> {
+                            product.sku == searchQuery
+                        }
 
-                    SkuSearchOptions.PartialMatch -> {
-                        searchQuery.let { query ->
-                            product.sku.contains(query)
-                        }
-                    }
-                }
-            }
-        }
+                        SkuSearchOptions.PartialMatch -> {
+                            searchQuery.let { query ->
+                                product.sku.contains(query)
+                            }
+                        }
+                    }
+                }
+            }
 
         return products
     }
Index: libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt
--- a/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt	(revision 07324874484fa97d2514918ad6b7166af9cd2414)
+++ b/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductsDao.kt	(date 1745347757473)
@@ -8,9 +8,8 @@
 
 @Dao
 internal abstract class ProductsDao {
-
-    @Query(
-        """
+    companion object {
+        private const val DEFAULT_QUERY = """
             SELECT * FROM ProductEntity
             WHERE localSiteId = :localSiteId
             AND (:status IS NULL OR status = :status)
@@ -21,7 +20,9 @@
             AND (remoteId NOT IN (:excludedProductIds))
             LIMIT CASE WHEN :limit IS NULL THEN -1 ELSE :limit END
         """
-    )
+    }
+
+    @Query(DEFAULT_QUERY)
     @Suppress("LongParameterList")
     abstract fun observeProducts(
         localSiteId: Int,
@@ -34,6 +35,19 @@
         excludedProductIds: List<Long>
     ): Flow<List<WCProductModel>>
 
+    @Query(DEFAULT_QUERY)
+    @Suppress("LongParameterList")
+    abstract suspend fun getProducts(
+        localSiteId: Int,
+        status: String?,
+        stockStatus: String?,
+        type: String?,
+        category: String?,
+        excludeSampleProducts: Boolean,
+        limit: Int?,
+        excludedProductIds: List<Long>
+    ): List<WCProductModel>
+
     @Query(
         """
             SELECT * FROM ProductEntity
```
</details>

WDYT?

wzieba replied:
Thanks for the suggestion, I applied it in e9736e125c8100a6732a6821a85620dd322b0b1e

My personal preference is to keep everything observable by default in DAOs to promote observability in the app. But I also understand the preference of readability with `suspend fun` for one-offs

hichamboushaba reviewed (commented):
Thanks so much @wzieba for the work on this, really great start 👏.

I didn't do a proper testing, I tested just few scenarios and reviewed some parts of the code.

I left some comments, some are blocking issues, so please take a look 🙏.

kidinov reviewed (commented):
Thank you @wzieba for your work here!

I left a couple of comments and tested the app - seems all good. There are a few good points from Hicham though!

wzieba committed:
Revert wrapping `loadProducts` else condition with `launch`

It's not necessary at this moment as `getProductList` is not `suspend` for now

wzieba committed:
Do not create index on `WCProductModel`

It's redundant: https://github.com/woocommerce/woocommerce-android/pull/13935/files?diff=split&w=0#r2054357967

wzieba committed:
Fix "exclude sample products" DAO condition

https://github.com/woocommerce/woocommerce-android/pull/13935#discussion_r2054067387

wzieba committed:
Move sorting back to database-level

https://github.com/woocommerce/woocommerce-android/pull/13935#discussion_r2054675234

wzieba committed:
Move searching back to db-level

Split `searchCachedProducts` from `getProducts` method in `WCProductStore` for improving clarity

https://github.com/woocommerce/woocommerce-android/pull/13935#discussion_r2054079848

wzieba committed:
Move counting products to db-level

This improves performance. Also, we can remove some of the parameters that didn't have use in scope of counting (like sorting or limit.

wzieba committed:
Introduce `getProducts` method to return filtered results via `suspend` fun instead of a `Flow`

wzieba committed:
Suppress `LongParameterList` in `ProductsDao`

We have to have long parameter names there.

wzieba commented:
Thank you for suggestions @kidinov @hichamboushaba ! 🙌  The PR is ready for a second round

wzieba committed:
Ignore case when ordering by name

Test for this was added in 86fd5a8a0e6a41db1f215c8608631de31b42db0d

GH discussion: https://github.com/woocommerce/woocommerce-android/pull/13943#pullrequestreview-2790671495

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/list/ProductListRepository.kt:
```diff
@@ -244,7 +244,7 @@ class ProductListRepository @Inject constructor(
     fun observeProductsCount(
         filterOptions: Map<WCProductStore.ProductFilterOption, String>,
-        excludeSampleProducts: Boolean = false
+        excludeSampleProducts: Boolean
```
Is this needed for the migration? I don't think it's, and personally I think it's safer to limit the PR to the migration related changes, WDYT?

And on another side, I think defaulting to false here was a conscious choice when we worked on the Blaze feature, as by default and for most features there is no need to consider whether the product is a sample or not, only a limited set of features would need to exclude them.

wzieba replied:
No, it's not needed but this method is used in one place only and the parameter is always `true` - I didn't really understand necessity of default `false` here.

But if this is something we want to keep, I don't have anything against: reverted in ce8a9135496be9236fe8635be3d16a86f28aacbe

hichamboushaba reviewed (approved):
Really great work @wzieba 👏

All looks good now, I left one non-blocking comment, but it's very minor.

wzieba committed:
Revert removing default argument from `ProductListRepository#observeProductsCount`

https://github.com/woocommerce/woocommerce-android/pull/13935/files/07324874484fa97d2514918ad6b7166af9cd2414..f2d08a9b87a3c33d4cbbe438aa8baed98985d203#r2058598381
